python/ray/tune/tests/test_tune_save_restore.py

```diff
@@ -0,0 +1,140 @@
+        shutil.rmtree(self.absolute_local_dir)
+        _register_all()
+
+    def _get_trial_dir(self, exp_dir):
+        for i in os.listdir(exp_dir):
+            if i.startswith("PG") and os.path.isdir(os.path.join(exp_dir, i)):
+                return i, os.path.join(exp_dir, i)
+        return None, None
+
+    def _train(self, exp_name, local_dir, absolute_local_dir):
+        self.absolute_local_dir = absolute_local_dir
+
+        trial, = tune.run(
+            "PG",
+            name=exp_name,
+            stop={"training_iteration": 1},
+            checkpoint_freq=1,
+            local_dir=local_dir,
+            config={
+                "env": "CartPole-v0",
+                "log_level": "FATAL"
+            })
+
+        exp_dir = os.path.join(absolute_local_dir, exp_name)
+        _, abs_trial_dir = self._get_trial_dir(exp_dir)
+
+        self.assertIsNone(trial.error_file)
+        self.assertEqual(trial.local_dir, exp_dir)
+        self.assertEqual(trial.logdir, abs_trial_dir)
+
+        self.assertTrue(os.path.isdir(absolute_local_dir), absolute_local_dir)
+        self.assertTrue(os.path.isdir(exp_dir))
+        self.assertTrue(os.path.isdir(abs_trial_dir))
+        self.assertTrue(
+            os.path.isfile(
+                os.path.join(abs_trial_dir, "checkpoint_1/checkpoint-1")))
+        os.chdir(self.current_dir)
+
+    def _restore(self, exp_name, local_dir, absolute_local_dir):
+        trial_name, abs_trial_dir = self._get_trial_dir(
+            os.path.join(absolute_local_dir, exp_name))
+
+        checkpoint_path = os.path.join(
+            local_dir, exp_name, trial_name,
+            "checkpoint_1/checkpoint-1")  # Relative checkpoint path
+
+        # The file tune would found. The absolute checkpoint path
+        tune_found_file = os.path.abspath(os.path.expanduser(checkpoint_path))
+        self.assertTrue(
+            os.path.isfile(tune_found_file),
+            "{} is not exist!".format(tune_found_file))
+
+        trial, = tune.run(
+            "PG",
+            name=exp_name,
+            stop={"training_iteration": 2},  # train one more iteration.
+            checkpoint_freq=1,
+            restore=checkpoint_path,  # Restore the checkpoint
+            max_failures=0,
+            config={
+                "env": "CartPole-v0",
+                "log_level": "FATAL"
+            },
+        )
+        self.assertIsNone(trial.error_file)
+
+    def testDottedRelativePath(self):
+        local_dir = "./test_dotted_relative_local_dir"
+        exp_name = "DottedRelativeLocalDir"
+        absolute_local_dir = os.path.abspath(local_dir)
+        self._train(exp_name, local_dir, absolute_local_dir)
+        self._restore(exp_name, local_dir, absolute_local_dir)
+
+    def testRelativePath(self):
+        local_dir = "test_relative_local_dir"
+        exp_name = "RelativePath"
+        absolute_local_dir = os.path.abspath(local_dir)
+        self._train(exp_name, local_dir, absolute_local_dir)
```

Comment:
Maybe move the `os.chdir(self.current_dir)` from `_train` to between the `self._train` and `self._restore`. This would have to be duplicated for every test, but that way it's clearer that what happens is 1) we train, 2) the directory gets moved back to the original, 3) we restore.